lib/portlayer/storage/volume_cache.go

```diff
@@ -0,0 +1,144 @@
+// Copyright 2016 VMware, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package storage
+
+import (
+	"net/url"
+	"os"
+	"sync"
+
+	log "github.com/Sirupsen/logrus"
+	"golang.org/x/net/context"
+)
+
+// VolumeLookupCache caches Volume references to volumes in the system.
+type VolumeLookupCache struct {
+
+	// Maps IDs to Volumes.
+	//
+	// id -> Volume
+	vlc     map[string]Volume
+	vlcLock sync.Mutex
+
+	// The underlying data storage implementation
+	volumeStore VolumeStorer
+}
+
+func NewVolumeLookupCache(ctx context.Context, vs VolumeStorer) (*VolumeLookupCache, error) {
+	v := &VolumeLookupCache{
+		vlc:         make(map[string]Volume),
+		volumeStore: vs,
+	}
+
+	return v, v.rebuildCache(ctx)
+}
+
+func (v *VolumeLookupCache) VolumeCreate(ctx context.Context, ID string, store *url.URL, capacityMB uint64, info map[string][]byte) (*Volume, error) {
+	// check if it exists
+	vol, err := v.VolumeGet(ctx, ID)
+	if err != nil && err != os.ErrNotExist {
+		return nil, err
+	}
+
+	// found in cache.
+	if vol != nil {
+		return vol, nil
+	}
+
+	vol, err = v.volumeStore.VolumeCreate(ctx, ID, store, capacityMB, info)
+	if err != nil {
+		return nil, err
+	}
+
+	v.vlcLock.Lock()
+	defer v.vlcLock.Unlock()
+
+	// Add it to the cache.
+	v.vlc[vol.ID] = *vol
+
+	return vol, nil
+}
+
+func (v *VolumeLookupCache) VolumeDestroy(ctx context.Context, ID string) error {
+	// Check if it exists
+	vol, err := v.VolumeGet(ctx, ID)
+	if err != nil && err == os.ErrNotExist {
+		return err
+	}
+
+	// remove it from the volumestore
+	if err = v.volumeStore.VolumeDestroy(ctx, ID); err != nil {
+		return err
+	}
+
+	v.vlcLock.Lock()
+	defer v.vlcLock.Unlock()
+	delete(v.vlc, vol.ID)
+
+	return nil
+}
+
+func (v *VolumeLookupCache) VolumeGet(ctx context.Context, ID string) (*Volume, error) {
+	v.vlcLock.Lock()
+	defer v.vlcLock.Unlock()
+
+	// look in the cache
+	vol, ok := v.vlc[ID]
+	if !ok {
+		return nil, os.ErrNotExist
+	}
+
+	return &vol, nil
+}
+
+func (v *VolumeLookupCache) VolumesList(ctx context.Context) ([]*Volume, error) {
+	v.vlcLock.Lock()
+	defer v.vlcLock.Unlock()
+
+	// look in the cache, return the list
+	var i int
+	l := make([]*Volume, len(v.vlc))
+	for _, vol := range v.vlc {
+		// this is idiotic
```

Comment:
Could just return a '[]Volume' instead?  In any case, can you use the range index (currently ignored '_'), rather than the 'var i' and 'i++' ?  Or even drop the index and use 'l = append(l, &e)'